### PR TITLE
cast function support

### PR DESCRIPTION
**Pull request by bbbearxyz**
1. bool to char/varchar
2. date to datetime
3. datetime to date
4. date to char/varchar
5. datetime to char/varchar

**What type of PR is this?**

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

**Which issue(s) this PR fixes:**

issue #2990

**What this PR does / why we need it:**

Not Available

**Special notes for your reviewer:**

Not Available

**Additional documentation (e.g. design docs, usage docs, etc.):**

Not Available

### CODE REVIEW

File: pkg/vectorize/typecast/typecast.go [hunk 1]
```diff
@@ -243,3 +248,53 @@ func UintToDecimal128[T constraints.Integer](xs []T, rs []types.Decimal128) ([]t
...
+	}
+	return rs, nil
+}
+
+func datetimeToBytes(xs []types.Datetime, rs *types.Bytes) (*types.Bytes, error) {
+	oldLen := uint32(0)
+	for _, x := range xs {
+		rs.Data = append(rs.Data, []byte(x.String())...)
+		newLen := uint32(len(rs.Data))
+		rs.Offsets = append(rs.Offsets, oldLen)
```
Comment (broccoliSpicy) [inline pkg/vectorize/typecast/typecast.go:295]: col := &types.Bytes{
 		Data:    make([]byte, 0, len(lvs)),
 		Offsets: make([]uint32, 0, len(lvs)),
 		Lengths: make([]uint32, 0, len(lvs)),
 	}


I think use append may cause problems here,

File: pkg/sql/plan2/function/operator/cast.go [hunk 2]
```diff
@@ -1010,6 +1106,57 @@ func CastStringAsDecimal128(lv, rv *vector.Vector, proc *process.Process) (*vect
...
+	lvs := lv.Col.([]types.Datetime)
+	col := &types.Bytes{
+		Data:    make([]byte, 0, len(lvs)),
+		Offsets: make([]uint32, 0, len(lvs)),
+		Lengths: make([]uint32, 0, len(lvs)),
+	}
+	if col, err = typecast.DateTimeToBytes(lvs, col); err != nil {
+		return nil, err
+	}
+	if err = proc.Mp.Gm.Alloc(int64(cap(col.Data))); err != nil {
```
Comment (broccoliSpicy) [inline pkg/sql/plan2/function/operator/cast.go:1120]: what is this for?

**Review by yingfeng — CHANGES_REQUESTED** [review]
conflict

**Review by broccoliSpicy — APPROVED** [review]

File: pkg/sql/plan2/function/operator/cast.go (hunk 2 above)
Comment (bbbearxyz) [reply pkg/sql/plan2/function/operator/cast.go:1120]: I just cal the memory cost in the process

**Pull request merged**